Simplenote/src/main/java/com/automattic/simplenote/viewmodels/TagDialogViewModel.kt

```diff
@@ -0,0 +1,98 @@
+package com.automattic.simplenote.viewmodels
+
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import com.automattic.simplenote.R
+import com.automattic.simplenote.analytics.AnalyticsTracker
+import com.automattic.simplenote.models.Tag
+import com.automattic.simplenote.repositories.TagsRepository
+
+class TagDialogViewModel(private val tagsRepository: TagsRepository) : ViewModel() {
+    private val _uiState = MutableLiveData<UiState>()
+    val uiState: LiveData<UiState> = _uiState
+
+    private val _event = SingleLiveEvent<TagDialogEvent>()
+    val event: LiveData<TagDialogEvent> = _event
+
+    fun start(tag: Tag) {
+        _uiState.value = UiState(tag.name, tag)
+    }
+
+    fun close() {
+        _event.postValue(CloseEvent)
+    }
+
+    fun updateUiState(tagName: String) {
+        if (tagName.isEmpty()) {
+            _uiState.value = _uiState.value!!.copy(tagName = tagName, errorMsg = R.string.tag_error_empty)
```

Comment:
Yes,every time I use `!!`, I do not feel good. There should be an invariant that when `updateUiState` is called, `_uiState.value` is never `null`. However, the compiler does not see this case. I have added a defensive check for the invariant that shows an error in case the invariant is not satisfies and hints the type system that the value is not null (Fixed in 1c0ef7c43e46b831b05cc3f8f73f9c58c52d2f39).

I also fixed the tests in 47f23d7ce08e1746b19c32b4e052e9e5d6eb1533.

Reply:
> Yes,every time I use !!, I do not feel good. 

Yes, I know, right! 😅 

>  There should be an invariant that when updateUiState is called, _uiState.value is never null. However, the compiler does not see this case. I have added a defensive check for the invariant that shows an error in case the invariant is not satisfies and hints the type system that the value is not null.

I like this, and it seems consistent with the rest.

> I also fixed the tests.

Actually, I think there is still one `!!` that can be converted to `?` in tests, is it not?

Reply:
The extra `!!` was fixed in bbfc4c1f7e8385de71a4f9e08f6655d173824141. 🙏